src/sentry/api/bases/external_actor.py

```diff
@@ -0,0 +1,128 @@
+from typing import Any, MutableMapping
+
+from django.db import IntegrityError
+from django.http import Http404
+from rest_framework import serializers
+from rest_framework.exceptions import PermissionDenied
+
+from sentry import features
+from sentry.api.serializers.rest_framework.base import CamelSnakeModelSerializer
+from sentry.api.validators.integrations import validate_provider
+from sentry.models import ExternalActor, Organization, OrganizationMember, Team, User
+from sentry.types.integrations import ExternalProviders, get_provider_choices
+
+AVAILABLE_PROVIDERS = {
+    ExternalProviders.GITHUB,
+    ExternalProviders.GITLAB,
+    ExternalProviders.SLACK,
+}
+
+
+class ExternalActorSerializerBase(CamelSnakeModelSerializer):  # type: ignore
+    external_id = serializers.CharField()
+    external_name = serializers.CharField(required=True)
+    provider = serializers.ChoiceField(choices=get_provider_choices(AVAILABLE_PROVIDERS))
+
+    @property
+    def organization(self):
+        return self.context["organization"]
+
+    def get_actor_id(self, validated_data: MutableMapping[str, Any]) -> int:
+        target_option = validated_data.pop(self._actor_key, None)
+
+        if not target_option:
+            raise Exception("Invalid actor")
```

Comment:
@mgaeta when would we get to this point if we have already validated the data? Is this error going to get surface to the user?

Reply:
You're right, this should never happen with validated data. I need to do the check to satisfy the type checker.

Reply:
Would not having `, None)` on the pop help fix this?